torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1840,6 +1850,82 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
       profilingTitle);
 }
 
+c10::intrusive_ptr<Work> ProcessGroupNCCL::allreduce_sparse(
+    std::vector<at::Tensor>& tensors,
+    const AllreduceOptions& opts) {
+#ifdef IS_NCCL_EXP
```

Comment:
Can we consider moving most part of this PR into the fb folder?
Its use of `#ifdef IS_NCCL_EXP` guard indicates that it is based on a NCCL version that does not publicly exist. fb folder would be the best place to host this kind of extension.

Reply:
Yeah, i agree it would be ideal to have this in the fb-only folder. There are also changes to the DDP reducer that should be specific to a certain nccl version, but that may be more challenging to split up.

Since there currently isnt a fb-specific ProcessGroupNccl, there will be additional work for BUCK targets / build rules to split the ProcessGroupNCCL.cpp source files into internal / external implementations. The PR is time sensitive so I will update this in a following PR

Reply:
Here is the initial diff https://www.internalfb.com/diff/D47068718, but it is not finished. The change is more involved than expected since there are a lot of common things shared and may warrant some design discussion on what parts of ProcessGroupNCCL to separate into its own file to share amongst internal and external